examples/mpm88.py

```diff
@@ -1,87 +1,87 @@
+            grid_m[base + offset] += weight * p_mass
+    for i, j in grid_m:
+        if grid_m[i, j] > 0:
+            grid_v[i, j] /= grid_m[i, j]
+        grid_v[i, j][1] -= dt * gravity
+        if i < bound and grid_v[i, j].x < 0:
+            grid_v[i, j].x = 0
+        if i > n_grid - bound and grid_v[i, j].x > 0:
+            grid_v[i, j].x = 0
+        if j < bound and grid_v[i, j].y < 0:
+            grid_v[i, j].y = 0
+        if j > n_grid - bound and grid_v[i, j].y > 0:
+            grid_v[i, j].y = 0
+    for p in x:
+        Xp = x[p] / dx
+        base = int(Xp - 0.5)
+        fx = Xp - base
+        w = [0.5 * (1.5 - fx)**2, 0.75 - (fx - 1)**2, 0.5 * (fx - 0.5)**2]
+        new_v = ti.Vector.zero(ti.f32, 2)
+        new_C = ti.Matrix.zero(ti.f32, 2, 2)
+        for i, j in ti.static(ti.ndrange(3, 3)):
+            offset = ti.Vector([i, j])
+            dpos = (offset - fx) * dx
+            weight = w[i].x * w[j].y
+            g_v = grid_v[base + offset]
+            new_v += weight * g_v
+            new_C += 4 * weight * g_v.outer_product(dpos) / dx**2
+        v[p] = new_v
+        x[p] += dt * v[p]
+        J[p] *= 1 + dt * new_C.trace()
+        C[p] = new_C
 
-  for i, j in grid_m:
-    if grid_m[i, j] > 0:
-      bound = 3
-      inv_m = 1 / grid_m[i, j]
-      grid_v[i, j] = inv_m * grid_v[i, j]
-      grid_v[i, j][1] -= dt * 9.8
-      if i < bound and grid_v[i, j][0] < 0:
-        grid_v[i, j][0] = 0
-      if i > n_grid - bound and grid_v[i, j][0] > 0:
-        grid_v[i, j][0] = 0
-      if j < bound and grid_v[i, j][1] < 0:
-        grid_v[i, j][1] = 0
-      if j > n_grid - bound and grid_v[i, j][1] > 0:
-        grid_v[i, j][1] = 0
-
-  for p in x:
-    base = (x[p] * inv_dx - 0.5).cast(int)
-    fx = x[p] * inv_dx - base.cast(float)
-    w = [
-        0.5 * (1.5 - fx) ** 2, 0.75 - (fx - 1.0) ** 2, 0.5 * (fx - 0.5) ** 2
-    ]
-    new_v = ti.Vector.zero(ti.f32, 2)
-    new_C = ti.Matrix.zero(ti.f32, 2, 2)
-    for i in ti.static(range(3)):
-      for j in ti.static(range(3)):
-        dpos = ti.Vector([i, j]).cast(float) - fx
-        g_v = grid_v[base + ti.Vector([i, j])]
-        weight = w[i][0] * w[j][1]
-        new_v += weight * g_v
-        new_C += 4 * weight * g_v.outer_product(dpos) * inv_dx
-    v[p] = new_v
-    x[p] += dt * v[p]
-    J[p] *= 1 + dt * new_C.trace()
-    C[p] = new_C
-
-for i in range(n_particles):
-  x[i] = [random.random() * 0.4 + 0.2, random.random() * 0.4 + 0.2]
-  v[i] = [0, -1]
-  J[i] = 1
+@ti.kernel
+def init():
+    for i in range(n_particles):
+        x[i] = [ti.random() * 0.4 + 0.2, ti.random() * 0.4 + 0.2]
+        v[i] = [0, -1]
+        J[i] = 1
 
-gui = ti.GUI("MPM88", (512, 512))
-for frame in range(20000):
+init()
+gui = ti.GUI('MPM88')
+while gui.running and not gui.get_event(gui.PRESS):
   for s in range(50):
-    grid_v.fill([0, 0])
-    grid_m.fill(0)
     substep()
   gui.clear(0x112F41)
   gui.circles(x.to_numpy(), radius=1.5, color=0x068587)
```

Comment:
It seems that the indentation of this part is inconsistent with the rest of the code?